File: Sliding Window/209. Minimum Size Subarray Sum/209. Minimum Size Subarray Sum.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function (target, nums) {
  let start = 0;
  let summ = 0;
  let min_length = Infinity;

  for (let i = 0; i < nums.length; i++) {
    summ += nums[i];

    while (summ >= target) {
      min_length = Math.min(min_length, i - start + 1);
      summ -= nums[start];
      start++;
    }
  }

  if (min_length === Infinity) return 0;
  return min_length;
};

// Time: O(n)
// Space: O(1)

File: Sliding Window/209. Minimum Size Subarray Sum/209. Minimum Size Subarray Sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        start = 0
        summ = 0
        min_length = float("inf")

        for i in range(len(nums)):
            summ += nums[i]
            
            while summ >= target:
                min_length = min(min_length, i - start + 1)
                summ -= nums[start]
                start += 1
        
        return min_length if min_length != float("inf") else 0

# Time: O(n)
# Space: O(1)